File: Recursion/Lecture 08/LeetCode/Permutations II.py
"""
This question is exactly similar to Permutations - I 
Only difference being we have duplicate elements in nums[]. But we need to print Unique Permutations only
For that purpose, we have used a local unordered_set data structure which tells us if that particular element has been previosuly encountered or not
If we have taken / chosen the element before, we will not include it anymore. 
But if we haven't included the element, we will include it now
"""
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        
        def hashKey(arr):
            s = ""
            for i in arr:
                s += str(i)
            return s
            
        def helper(pos, n, key):
            
            #  Base Case executes when we have traversed the entire nums[]
            if pos >= n:
                ans.append(nums.copy())
                numsMap[key] = 1
                return
            
            #  The rest logic remains exactly the same
            
            for i in range(pos, n):
                
                # We simply use our swapping logic to create Permutations  
                
                nums[i], nums[pos] = nums[pos], nums[i]
                
                # hasing the nums arry to a string, to store in numsMap dictionary
                
                key = hashKey(nums)
                
                # And we ask Recursion to handle rest of the task
                # To ensure we're not storing any duplicate | we check if the hash(nums) is already not present in the numsMap
                if key not in numsMap:
                    helper(pos+1, n, key)
                
                # Backtrack and undo the change we have done 
                
                nums[i], nums[pos] = nums[pos], nums[i] 
            return
        
        ans = []
        
        # Dictionary to store each permutation of nums as key | before  storing it in key we need to convert nums [] to a String, as list are not hashable
        # Dictionary is used to optimise the Time complexit as search TC is O(1) 
        numsMap = {}
        
        helper(0, len(nums), hashKey(nums))
        
        return ans
            
"""
Time Complexity  : O(N * N!)
Space Complexity : O(N)
"""

File: Recursion/Lecture 09/LeetCode/Permutations II.java
/* The idea is to generate all Possible Permutations by swapping the nums[currentIndex] with nums[i]
As all the characters are Unique, the Permutations generated will also be Unique
We make neccessary changes in nums[] vector only. We don't use any other data structure
For Permutations, order doesn't matter

NOTE : In java there are no swap function so we need to create our own swap function

NOTE : The array must be converted into List before storing it in answer
*/


class Solution 
{
    public void helper(int pos , int n , int[] nums , List<List<Integer>> ans)
    {
        // If we have reached the end of nums, we have found a Valid Permutations of nums

        if(pos >= n)
        {
            List<Integer> permutation = ConverttoArray(nums);
            ans.add(new ArrayList(permutation)) ;
            return ;
        }

        // Otherwise we iterate over all the other elements and try to generate Permutations by swapping element at pos with element at i

        // Hashset ensures we are not taking duplicates and thus not making Duplicate Permutation
        
        HashSet<Integer> set = new HashSet<>();
        
        for(int i = pos ; i < n ; i++)
        {
            // If we have encountered the element before, we will simply skip the rest of iterations
            
            if(set.contains( nums[i] ))
                continue;
            
             // We insert nums[i] so that we don't create Duplicate Permutations
            
            set.add(nums[i]);
            
            // We create one Unique Permutation by swapping

            swap(pos , i, nums) ;

            // And we ask Recursion to handle rest of the task

            helper(pos + 1 , n , nums , ans) ;

            // But after we come back, we must backtrack and undo the changes we have done

            swap(pos , i, nums) ;
        }

        return ;
    }
    
    public void swap (int i, int j, int[] nums)
    {
        // Swaping the elements
        
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    
    public List<Integer> ConverttoArray(int[] nums)
    {
        // Converting the array into List
        
        List<Integer> ans = new ArrayList<>();
        for(int ele : nums)
            ans.add(ele);
        return ans;
    }
    
    public List<List<Integer>> permuteUnique(int[] nums) 
    {
        List<List<Integer>> ans = new ArrayList<>();

        helper(0 , nums.length , nums , ans) ;

        return ans ;
    }
}

/*
Time Complexity  : O(N * N!)
Space Complexity : O(N)
*/
